### PR TITLE
rose_prune: can now prune any item with a cycle point in its path name

### LINKED ISSUE #1418: rose_prune archive-logs-at is broken by some cycle point formats
Cycle points of the format ```CCYY-MM-DD-Thh``` do not match the regular expression defined at  ```REC_CYCLE_TIME``` at ```lib/python/rose/suite_engine_procs/cylc.py:93```, for instance.

### LINKED ISSUE #1491: rose_prune - missing features?
I'm just putting these up for discussion; it could potentially wait for cylc/cylc#994, cylc/cylc#1159
 
- [x] deletion of log tarballs at some offset
- [x] keeping all output under the datac directories is an admirable goal, to make housekeeping simple, but sometimes it's not convenient.  E.g. for a suite sub-system that continually reprocesses the previous n months data, this data would likely (a) need different housekeeping from the rest of the suite output; and (b) be kept in a flat directory with cycle point filenames. So, it would be good if rose_prune could also house-keep
   * non-datac directory locations
   * cycle point labelled filenames
- [x] NOT globbing.  This comment suggests it can be done already: https://github.com/metomi/rose/issues/898#issuecomment-30506941  - can it? The code seems to simply run ```ls``` and ```rm``` via the shell, on the glob pattern, but I need ```shopt -s extglob``` to get NOT globs in the shell.

### PR DESCRIPTION
**Pull request by matthewrmshin**
New `prune{ITEM}=CYCLE[:GLOBS] ...` allow prune of items under any
sub-directories. The old `prune-work-at` and `prune-datac-at` are now
deprecated aliases of `prune{work}` and `prune{sharecycle}`.

Cycle points can now be date-time points or offsets of the current cycle
point.

Each glob in the GLOBS string can now contain a `%(cycle)s`
substitution. When a glob is specified like so, the program will not add
the cycle under ITEM as a sub-directory, but will substitute `%(cycle)s`
in the glob with the cycle.

Now fail if a cycle point in the configuration has a bad syntax.

Close #1418. Close #1491.

### CODE REVIEW

**Comment by matthewrmshin** [comment]
N.B. This change will make more sense after #1571 is merged.

**Comment by matthewrmshin** [comment]
@arjclark please review.

File: doc/rose-rug-task-run.html [hunk 1]
```diff
@@ -605,9 +605,10 @@ <h2 id="rose-task-run.util.rose_prune">Built-in Application:
     in a <var>rose-app.conf</var>.</p>
 
     <p>All settings are expressed as a space delimited list of cycles, normally
-    as date/time offsets relative to the current cycle, in a format recognised
-    by the <code>--offset=OFFSET</code> option of <code>rose date</code>. E.g.
-    <kbd>-P1DT6H</kbd> is 1 day and 6 hours before the current cycle time.</p>
+    as cycle points or a offsets relative to the current cycle. For date time
```
Comment (arjclark) [inline doc/rose-rug-task-run.html]: Either "or *as* offsets" or "or ~~a~~ offsets"

Comment (arjclark) [inline doc/rose-rug-task-run.html]: Should that be For "date-time" cycles ?

File: doc/rose-rug-task-run.html [hunk 2]
```diff
@@ -633,27 +634,39 @@ <h2 id="rose-task-run.util.rose_prune">Built-in Application:
...
+        clear out <var>work/20141225T0000Z/</var>.
 
         <p>If <var>globs</var> are specified for a cycle, it will attempt to
-        prune only items matching <var>work/CYCLE/GLOBS</var>. E.g. In cylc,
-        if current cycle is <samp>20141225T1200Z</samp>,
-        <kbd>prune-work-at=-PT12H:wild*</kbd> will clear out all items matching
-        <var>work/20141225T0000Z/wild*</var>.</p>
+        prune only items matching <var>CYCLE/GLOBS</var> under
+        <var>item-root</var>. E.g. In cylc, if current cycle is
+        <samp>20141225T1200Z</samp>, <kbd>prune{share/cycle}=-PT12H:wild*</kbd>
```
Comment (arjclark) [inline doc/rose-rug-task-run.html]: ```<samp>20141225T1200Z</samp>```, *then*  ```<kbd>prune{share/cycle}=-PT12H:wild*</kbd>```

File: doc/rose-rug-task-run.html [hunk 3]
```diff
@@ -633,27 +634,39 @@ <h2 id="rose-task-run.util.rose_prune">Built-in Application:
...
+        will clear out all items matching
+        <var>share/cycle/20141225T0000Z/wild*</var>.</p>
+
+        <p>A glob can also be specified as a formatting string containing a
+        single substitution <code>%(cycle)s</code>. In this mode, the cycle
+        string will not be added as a sub-directory of the item-root. E.g. In
+        cylc, if current cycle is <samp>20141225T1200Z</samp>,
```
Comment (arjclark) [inline doc/rose-rug-task-run.html]: , then

File: doc/rose-rug-task-run.html [hunk 4]
```diff
@@ -633,27 +634,39 @@ <h2 id="rose-task-run.util.rose_prune">Built-in Application:
...
+        <kbd>prune{share}=-PT12H:hello-*-at-%(cycle)s.txt</kbd> will clear
+        out all items matching
+        <var>share/hello-*-at-20141225T0000Z.txt</var>.</p>
       </dd>
 
+      <dt><kbd>prune-work-at=cycle[:globs] ...</kbd></dt>
```
Comment (arjclark) [inline doc/rose-rug-task-run.html]: Perhaps lead with new syntax and something like "replaces deprecated ..."

File: doc/rose-rug-task-run.html [hunk 5]
```diff
@@ -633,27 +634,39 @@ <h2 id="rose-task-run.util.rose_prune">Built-in Application:
...
+      <dd><em>Deprecated.</em> Equivalent to
+      <kbd>prune{work}=cycle[:globs] ...</kbd></dd>
+
       <dt><kbd>prune-datac-at=cycle[:globs] ...</kbd></dt>
 
-      <dd>Remove the <var>ROSE_DATAC</var> of the specified cycles. If
-      <var>globs</var> are specified for a cycle, only items matching the globs
-      in the <var>ROSE_DATAC</var> directories of the specified cycle will be
-      pruned.</dd>
+      <dd><em>Deprecated.</em> Equivalent to
```
Comment (arjclark) [inline doc/rose-rug-task-run.html]: As above.

**Comment by matthewrmshin** [comment]
Fixed docs.

**Comment by arjclark** [comment]
Generally looks good, some minor tweaks to be made to the documentation.

Am getting a reproducible failure in my environment for one of the tests as:
```
rose test-battery t/rose-task-run/06-app-prune-iso.t
--- edited-prune.log	2015-04-24 14:40:31.726150409 +0100
+++ expected-prune.log	2015-04-24 14:40:31.728150445 +0100
@@ -1,4 +1,3 @@
-ycle/d;
 [INFO] delete: hpc2e-l001:log/job/20130101T0000Z/
 [INFO] delete: hpc2e-l001:share/cycle/20130101T0000Z/a-hello-2
 [INFO] delete: hpc2e-l001:work/20130101T0000Z
t/rose-task-run/06-app-prune-iso.t .. Failed 1/7 subtests 

Test Summary Report
-------------------
t/rose-task-run/06-app-prune-iso.t (Wstat: 0 Tests: 7 Failed: 1)
  Failed test:  1
Files=1, Tests=7, 76 wallclock secs ( 0.02 usr  0.01 sys + 20.42 cusr  7.40 csys = 27.85 CPU)
```

**Comment by arjclark** [comment]
Looks good to me. Test now working in my environment.

**Pull request merged**